src/plugins/macaddr/macaddr.c

```diff
@@ -0,0 +1,271 @@
+	}
+}
+
+void transformToInt (Key * key)
+{
+	const char * mac = keyString (key);
+
+	char * macReduced = elektraMalloc (12);
+
+	removeStandardizedSeparator (mac, macReduced);
+
+	long intValue = strtol (macReduced, NULL, 16);
+
+	const int n = snprintf (NULL, 0, "%lu", intValue);
+	char * buffer = elektraMalloc (n + 1);
+	snprintf (buffer, n + 1, "%lu", intValue);
+
+	keySetString (key, buffer);
+	elektraFree (macReduced);
+	elektraFree (buffer);
+}
+
+int elektraMacaddrGet (Plugin * handle ELEKTRA_UNUSED, KeySet * returned, Key * parentKey)
+{
+	if (!elektraStrCmp (keyName (parentKey), "system/elektra/modules/macaddr"))
+	{
+		KeySet * contract =
+			ksNew (30, keyNew ("system/elektra/modules/macaddr", KEY_VALUE, "macaddr plugin waits for your orders", KEY_END),
+			       keyNew ("system/elektra/modules/macaddr/exports", KEY_END),
+			       keyNew ("system/elektra/modules/macaddr/exports/get", KEY_FUNC, elektraMacaddrGet, KEY_END),
+			       keyNew ("system/elektra/modules/macaddr/exports/set", KEY_FUNC, elektraMacaddrSet, KEY_END),
+
+#include ELEKTRA_README
+
+			       keyNew ("system/elektra/modules/macaddr/infos/version", KEY_VALUE, PLUGINVERSION, KEY_END), KS_END);
+		ksAppend (returned, contract);
+		ksDel (contract);
+
+		return ELEKTRA_PLUGIN_STATUS_SUCCESS;
+	}
+	ksRewind (returned);
+	Key * cur;
+	while ((cur = ksNext (returned)) != NULL)
+	{
+		const Key * meta = keyGetMeta (cur, "check/macaddr");
+		if (!meta) continue;
+		const Key * origValue = keyGetMeta (cur, "origvalue");
+		if (origValue)
+		{
+			ELEKTRA_SET_ERRORF (
+				ELEKTRA_ERROR_STATE, parentKey,
+				"Meta key 'origvalue' for key %s not expected to be set, another plugin has already set this meta key!",
+				keyString (cur));
+			return ELEKTRA_PLUGIN_STATUS_ERROR;
+		}
+		int rc = validateMac (cur);
+		if (rc)
+		{
+			ELEKTRA_SET_ERRORF (ELEKTRA_ERROR_INVALID_FORMAT, parentKey, "%s is not in a supported format.", keyString (cur));
+			return ELEKTRA_PLUGIN_STATUS_ERROR;
+		}
+
+		if (checkIntMac (keyString (cur)))
+		{
+			transformMac (cur);
+			transformToInt (cur);
+		}
+		return ELEKTRA_PLUGIN_STATUS_SUCCESS;
+	}
+
+	return ELEKTRA_PLUGIN_STATUS_NO_UPDATE;
+}
+
+int elektraMacaddrSet (Plugin * handle ELEKTRA_UNUSED, KeySet * returned ELEKTRA_UNUSED, Key * parentKey ELEKTRA_UNUSED)
+{
+	ksRewind (returned);
+	Key * cur;
+	while ((cur = ksNext (returned)) != NULL)
+	{
+		const Key * meta = keyGetMeta (cur, "check/macaddr");
+		if (!meta) continue;
+		int rc = validateMac (cur);
+		if (rc)
+		{
+			ELEKTRA_SET_ERRORF (ELEKTRA_ERROR_INVALID_FORMAT, parentKey, "%s is not in a supported format.", keyString (cur));
+			return ELEKTRA_PLUGIN_STATUS_ERROR;
+		}
+		const Key * origValue = keyGetMeta (cur, "origvalue");
+		if (origValue)
+		{
+			keySetString (cur, keyString (origValue));
+			return ELEKTRA_PLUGIN_STATUS_SUCCESS;
+		}
+		keySetMeta (cur, "origvalue", keyString (cur));
+
+		if (checkIntMac (keyString (cur)))
+		{
+			transformMac (cur);
```

Comment:
You shouldn't transform in `kdbSet`. If `origvalue` is present, restore the value. Then validate the value (whether restored or not) and store it. This ensures that we always use the format that the user chose when setting a value.